api/rosetta/balance.go

```diff
@@ -42,35 +44,79 @@ func (d *Data) Balance(ctx echo.Context) error {
 	var req BalanceRequest
 	err := ctx.Bind(&req)
 	if err != nil {
-		return echo.NewHTTPError(http.StatusBadRequest, object.AnyError(err))
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat(err.Error()))
 	}
 
-	err = d.validate.Network(req.NetworkID)
-	if err != nil {
-		return echo.NewHTTPError(http.StatusUnprocessableEntity, object.AnyError(err))
+	if req.NetworkID.Blockchain == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("blockchain identifier blockchain missing"))
 	}
-	err = d.validate.Block(req.BlockID)
-	if err != nil {
-		return echo.NewHTTPError(http.StatusUnprocessableEntity, object.AnyError(err))
+	if req.NetworkID.Network == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("blockchain identifier network missing"))
 	}
-	err = d.validate.Account(req.AccountID)
-	if err != nil {
-		return echo.NewHTTPError(http.StatusUnprocessableEntity, object.AnyError(err))
+
+	if req.BlockID.Index == 0 && req.BlockID.Hash == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("block identifier at least one of hash or index"))
+	}
+	if req.BlockID.Hash != "" && len(req.BlockID.Hash) != hexIDSize {
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("block identifier hash wrong length (have: %d, want: %d)", len(req.BlockID.Hash), hexIDSize))
+	}
+
+	if req.AccountID.Address == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("account identifier address missing"))
+	}
+	if len(req.AccountID.Address) != hexAddressSize {
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("account identifier address wrong length (have: %d, want: %d)", len(req.AccountID.Address), hexAddressSize))
+	}
+
+	if len(req.Currencies) == 0 {
+		return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("currency identifiers empty"))
 	}
 	for _, currency := range req.Currencies {
-		err = d.validate.Currency(currency)
-		if err != nil {
-			return echo.NewHTTPError(http.StatusUnprocessableEntity, object.AnyError(err))
+		if currency.Symbol == "" {
+			return echo.NewHTTPError(http.StatusBadRequest, InvalidFormat("currency identifier symbol missing"))
 		}
 	}
```

Comment:
I principle, I agree with the sentiment. However, I did not want to investigate this further, because I don't think the overhead in putting structure in place is worth the gained value. The controller only decodes, validates and encodes, so I think the responsibility is isolated enough.

I would prefer to have it a bit less redundant, but at least it is clear, explicit and easy to change (also unlikely to change!). So I will leave it like this for now, and create an issue for refactoring this into an echo middleware for validation: